components/channel/channel.h

```diff
@@ -0,0 +1,78 @@
+#ifndef DATADOG_PHP_CHANNEL_H
+#define DATADOG_PHP_CHANNEL_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+typedef struct datadog_php_receiver_s datadog_php_receiver;
+typedef struct datadog_php_sender_s datadog_php_sender;
+typedef struct datadog_php_channel_s datadog_php_channel;
+typedef struct datadog_php_channel_impl_s datadog_php_channel_impl;
+
+struct datadog_php_receiver_s {
+    /**
+     * Receives on item, waiting for a send to the channel if necessary.
+     * However, it may return empty-handed and return false, such as if there are
+     * no senders left, or the timeout is reached, or if it is interrupted.
+     */
+    bool (*recv)(datadog_php_receiver *self, void **data, uint64_t timeout_nanos);
+
+    /**
+     * Destructs the receiver.
+     */
+    void (*dtor)(datadog_php_receiver *self);
+
+    // private:
+    datadog_php_channel_impl *channel;
+};
+
+struct datadog_php_sender_s {
+    /**
+     * Sends data through the channel. May fail, in which case the caller is
+     * responsible for dtor'ing the data if applicable. Does not wait for the
+     * data to be received, only enqueued.
+     */
+    bool (*send)(datadog_php_sender *self, void *data);
+
+    /**
+     * Clones the sender. Can fail if there are too many senders already, in
+     * which false is returned and the caller should not dtor the clone.
+     */
+    bool (*clone)(datadog_php_sender *self, datadog_php_sender *clone);
```

Comment:
This doesn't appear to be used or tested, and the only thing it's supposed to do (enforce sender limit) is left as a todo.

Could you talk a little bit about why this is required, where it would be used, and include a test for it ?

Alternatively, if this doesn't look necessary right now, maybe drop it ?

Reply:
`clone` is supposed to be how you get an additional sender. For instance, if we had an HTTP uploader channel and the tracer and profiler both send HTTP messages then we'd have multiple senders (and still just one receiver, the thing that does the uploads).

At the moment this is unused, correct. I'll think about adding a test vs removing it for now.

Reply:
I have added a test for non-contended multiple-senders which uses `.clone`.